### ISSUE TITLE
Sortable items in admin UI always jump to the end

### ISSUE THREAD

**Issue by DHDaniel** [keystonejs/keystone#4651]
### Expected behavior
When setting `sortable: true` when creating a new model (`new keystone.List()`), it should allow you to drag and drop the items in the Admin UI to rearrange their `sortOrder` (as per the documentation at http://keystonejs.com/docs/database/#lists-options).

### Actual/Current behavior
When dragging an item and dropping it in between two other items, it does not take that place. Instead, it drops down to the end of the item 'queue', taking the last place in the sort order.

### Steps to reproduce the actual/current behavior
Enable the `sortable` option when creating a new `keystone.List()`, and then attempt to rearrange the sort order of items in the list via the Admin UI when there are more than three items (i.e attempting to place an item in between two items).

### Environment

| Software         | Version
| ---------------- | -------
| macOS High Sierra   | 10.13.4
| Keystone         | 4.0.0-beta.4
| Node               | v8.9.0
| Browser           | Chrome

*Note*: the Keystone version being used is not the most recent, as I haven't updated. However, when looking through the release notes of the subsequent versions, I haven't seen any bug fixes relating to this issue.

**Comment by autoboxer** [comment]
@DHDaniel, would you please test again with the latest version and let us know if this is still an issue?  While no bug fixes were listed for this in beta.5, beta.7, or beta.8, there were a large number of commits in the last few releases, and change history wasn't recorded until after the fact.  It's possible it was fixed and not recorded.

**Comment by DHDaniel** [comment]
@autoboxer sorry for not updating to beta.8 from the start, I was using Keystone for an app in production and was afraid any updates may break it (why fix something that isn't broken? Haha).

I've updated to v4.0.0-beta.8 and the problem still occurs. However, it now doesn't always jump to the end of the queue - instead, it randomly takes another position in the list. Attached are two photos demonstrating it.

**Image 1**: attempting to move 'DK Email' to position between 'Mobile phone 1' and 'LinkedIn Profile'.
**Image 2**: actual result.

<img width="1341" alt="Dragging item into place, about to drop." src="https://user-images.githubusercontent.com/16171992/40984597-48db6258-68b0-11e8-8ba3-294cf6d56863.png">
<img width="1341" alt="Position where item is dropped. The cursor is where it was supposed to be dropped." src="https://user-images.githubusercontent.com/16171992/40984599-4907af8e-68b0-11e8-9a79-5d37425a5550.png">

**Comment by laurenskling** [comment]
Adding sortable adds the sortOrder field to the item. If you add sortable to an existing model, the items in the database won't have this field (yet). Sorting an item will only change a selection of the items, not the entire collection. If items don't have sortOrder yet l, it won't be set. So adding sortable to an existing model will result in an endless misplacement of items, because of the missing sortOrder field.

**Comment by fogarasyroland** [comment]
I added sortable to existing model and added sortOrder to all existing docs too (with proper values), but I'm get same bug as described above

**Comment by laurenskling** [comment]
I just tested `sortable: true` again on a new (!) model, it works properly.
It's very sensitive when adding to an existing model. Every document needs the `sortOrder` field, with a correct (!) number starting at `1`. If you do not have a correct order, it will result in weird sorting, because only numbers that _need_ to change are changed (by adding or extracting 1), so setting everthing to `1` will not make it work. Numbers need to be correct.

If you are positive it's not working with a complete new model, please give me a test setup that proves it.

**Comment by laurenskling** [comment]
now I'm thinking of it... deleting a document will result in a broken sortOrder chain. The deleted number is missing. So after deleting a document, sortOrder will be buggy.....

having 5 items, deleting the 3rd, dragging the one that has `4` on the document to the bottom (which would be 4) results in it keeping `4` and 5 doesn't change, so it'll be stuck as the third in line.

**Comment by savager** [comment]
yes, this seems to be what I'm running into also. I just downloaded the production database my sortable items appear with a sortOrder of:  length:54     
[ 12,  
  13,  
  54,  
  59,  
  28,  
  52,  
  48,  
  7,   
  5,   
  19,  
  134, 
  135, 
  136, 
  137, 
  138, 
  139, 
  11,  
  140, 
  141, 
  142, 
  143, 
  144, 
  71,  
  6,   
  1,   
  14,  
  23,  
  25,  
  26,  
  27,  
  24,  
  21,  
  29,  
  30,  
  36,  
  3,   
  8,   
  9,   
  146, 
  4,   
  148, 
  149, 
  20,  
  150, 
  152, 
  154, 
  2,   
  156, 
  15,  
  159, 
  160, 
  10,  
  161, 
  162 ]

I suppose for now I'll write an update that manually restores normal sortOrder and make a note about removing any...

**Comment by savager** [comment]
So, just to follow up on this for anyone in the future I "solved" this by adding a post remove hook to my sortable content types that updated the sortOrder value when one is deleted. Then I added an update that manually reset the sortOrder on existing data. 

That said, I believe sortOrder should be completely disabled as it currently stands. This is a pretty glaring issue, and while filtered also doesn't work.

**Comment by laurenskling** [comment]
I agree it's pretty flacky. But I'm unsure if it's flacky by design or ended up this way. Maybe one of the original creators of this feature can shed a light on the matter?

**Comment by savager** [comment]
Yeah, I mean I see that there is abandoned code where someone made an effort to be able to drag and drop between pages. imho it's nowhere near that point of feature. I'd say the drag and drop needs to be refactored but I shouldn't talk because I don't have time for it.

**Comment by joephuz** [comment]
having this issue, can confirm it stems from deleting.
@savager - would you mind sharing your post remove hook, your wrote?

**Comment by joephuz** [comment]
Hey guys, Ive made a fix in the following fork... Not sure if its the most elegant but its working well for my project:
https://github.com/joephuz/keystone-classic/tree/bug-fix%2C-%234651

**Comment by internetErik** [comment]
For anyone looking for the solution that Savager was using, it may look something like this:

```JavaScript
ListName.schema.post('remove', () => {
  keystone.list('ListName')
  .model
  .find()
  .sort({ sortOrder: 1 })
  .exec((err, results) => {
    for(let i = 0; i < results.length; i++) {
      let item = results[i];
      if(item.sortOrder !== i+1) {
        item.sortOrder = i+1;
        item.save();
      }
    }
  });
});
```

Since you only have to run this on delete, it doesn't seem too high a cost to pay even when there are a lot of items you have to run through. I haven't tried this with deleting many items at a time
